python_ta/cfg/visitor.py

```diff
@@ -55,67 +84,72 @@ def visit_if(self, node: astroid.If) -> None:
             self._current_block = else_block
             for child in node.orelse:
                 child.accept(self)
             end_else = self._current_block
 
-        after_if_block = self.cfg.create_block()
-        self.cfg.link_or_merge(end_if, after_if_block)
-        self.cfg.link_or_merge(end_else, after_if_block)
+        after_if_block = self._current_cfg.create_block()
+        self._current_cfg.link_or_merge(end_if, after_if_block)
+        self._current_cfg.link_or_merge(end_else, after_if_block)
 
         self._current_block = after_if_block
 
     def visit_while(self, node: astroid.While) -> None:
         old_curr = self._current_block
 
         # Handle "test" block
-        test_block = self.cfg.create_block()
+        test_block = self._current_cfg.create_block()
         test_block.add_statement(node.test)
-        self.cfg.link_or_merge(old_curr, test_block)
+        self._current_cfg.link_or_merge(old_curr, test_block)
 
-        after_while_block = self.cfg.create_block()
+        after_while_block = self._current_cfg.create_block()
 
         # step into while
         self._control_boundaries.append((node, {astroid.Break.__name__: after_while_block,
                                                 astroid.Continue.__name__: test_block}))
 
         # Handle "body" branch
-        body_block = self.cfg.create_block(test_block)
+        body_block = self._current_cfg.create_block(test_block)
         self._current_block = body_block
         for child in node.body:
             child.accept(self)
         end_body = self._current_block
-        self.cfg.link_or_merge(end_body, test_block)
+        self._current_cfg.link_or_merge(end_body, test_block)
 
         # step out of while
         self._control_boundaries.pop()
 
         # Handle "else" branch
-        else_block = self.cfg.create_block(test_block)
+        else_block = self._current_cfg.create_block(test_block)
         self._current_block = else_block
         for child in node.orelse:
             child.accept(self)
         end_else = self._current_block
 
-        self.cfg.link_or_merge(end_else, after_while_block)
+        self._current_cfg.link_or_merge(end_else, after_while_block)
         self._current_block = after_while_block
 
     def visit_break(self, node: astroid.Break) -> None:
-        self._visit_continue_or_break(node)
+        self._visit_jump(node)
 
     def visit_continue(self, node:astroid.Continue) -> None:
-        self._visit_continue_or_break(node)
+        self._visit_jump(node)
+
+    def visit_return(self, node: astroid.Return) -> None:
+        self._visit_jump(node)
 
-    def _visit_continue_or_break(self, node: Union[astroid.Break, astroid.Continue]) -> None:
+    def _visit_jump(self, node: Union[astroid.Break, astroid.Continue, astroid.Return]) -> None:
         old_curr = self._current_block
+        req_boundary = astroid.FunctionDef if isinstance(node, astroid.Return) else astroid.While
```

Comment:
This refactoring is a great start, but awkward as the cases grow larger (e.g., with @dhslim's addition of support for `For` nodes).

But it turns out you don't need to actually check what the `req_boundary` needs to be at all. Instead, rely on the fact that the *second* element of the tuple is a dictionary mapping names of node classes to which block to connect to. So you just need to check whether `type(node).__name__ in exits` in the loop below.

(This begs the question of whether the first tuple element is necessary at all. I'm not sure, so we can leave it in for now.)